storagestate/storagestate.go

```diff
@@ -0,0 +1,730 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storagestate
+
+import (
+	"bytes"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aoscloud/aos_common/aoserrors"
+	"github.com/aoscloud/aos_common/aostypes"
+	"github.com/aoscloud/aos_common/api/cloudprotocol"
+	"github.com/aoscloud/aos_common/utils/fs"
+	"github.com/aoscloud/aos_communicationmanager/amqphandler"
+	"github.com/aoscloud/aos_communicationmanager/config"
+	"github.com/fsnotify/fsnotify"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/crypto/sha3"
+)
+
+/***********************************************************************************************************************
+ * Consts
+ **********************************************************************************************************************/
+
+const (
+	stateFileFormat  = "%s_state.dat"
+	stateChannelSize = 32
+)
+
+/***********************************************************************************************************************
+ * Vars
+ **********************************************************************************************************************/
+
+// These global variables are used to be able to mocking the functionality in tests.
+// nolint:gochecknoglobals
+var (
+	SetUserFSQuota     = fs.SetUserFSQuota
+	StateChangeTimeout = 1 * time.Second
+)
+
+// ErrNotExist is returned when requested entry not exist in DB.
+var (
+	ErrNotExist = errors.New("entry does not exist")
+	ErrNotFound = aoserrors.New("instance not found")
+)
+
+/***********************************************************************************************************************
+ * Types
+ **********************************************************************************************************************/
+
+// StorageStateInstanceInfo storage state instance info.
+type StorageStateInstanceInfo struct {
+	aostypes.InstanceIdent
+	InstancePath  string
+	StorageQuota  uint64
+	StateQuota    uint64
+	StateChecksum []byte
+}
+
+// MessageSender sends messages to the cloud.
+type MessageSender interface {
+	SendInstanceNewState(newState cloudprotocol.NewState) error
+	SendInstanceStateRequest(request cloudprotocol.StateRequest) error
+}
+
+// Storage storage interface.
+type Storage interface {
+	GetAllStorageStateInfo() ([]StorageStateInstanceInfo, error)
+	GetStorageStateInfo(instancePath string) (StorageStateInstanceInfo, error)
```

Comment:
aostypes.InstanceIdent is quite good to use as key value. It is comparable and more logical.
```go
GetStorageStateInfo(instanceIdent aostypes.InstanceIdent) (StorageStateInstanceInfo, error)
```

Reply:
Done